crmprtd/align.py

```diff
@@ -8,12 +8,222 @@
+    log.debug('Search for active stations')
+    for history in histories:
+        hist = sesh.query(History).filter(
+            and_(History.id == history.id,
+                 History.sdate != None,
+                 History.edate == None))    # noqa
+
+        if hist.count() == 1:
+            log.debug('Matched history',
+                      extra={'station_name': hist.first().station_name})
+            return hist.first()
+
+    log.warning('Could not find match',
+                extra={'network_name': obs_tuple.network_name,
+                       'native_id': obs_tuple.station_id})
+    return None
+
+
+def match_station_with_location(sesh, obs_tuple, histories):
+    log.debug('Find matching station with location')
+    close_stns = closest_stns_within_threshold(sesh, obs_tuple.lon,
+                                               obs_tuple.lat, 800)
+
+    if len(close_stns) == 0:
+        log.debug('No station nearby')
+        return create_station_and_history_entry(sesh, obs_tuple)
+
+    for id in close_stns:
+        for history in histories:
+            if id == history.id:
+                log.debug('Matched history',
+                          extra={'station_name': history.station_name})
+                return history
+
+
+def match_station(sesh, obs_tuple, history):
+    if obs_tuple.lat and obs_tuple.lon:
+        return match_station_with_location(sesh, obs_tuple, history)
+    else:
+        return match_station_with_active(sesh, obs_tuple, history)
+
+
+def create_station_and_history_entry(sesh, obs_tuple):
+    network = sesh.query(Network).filter(
+        Network.name == obs_tuple.network_name)
+
+    network = network.first()
+    stn = Station(native_id=obs_tuple.station_id, network=network)
+    with sesh.begin_nested():
+        sesh.add(stn)
+    log.info('Created new station entry',
+             extra={'native_id': stn.native_id, 'network_name': network.name})
+
+    hist = History(station=stn,
+                   lat=obs_tuple.lat,
+                   lon=obs_tuple.lon)
+
+    with sesh.begin_nested():
+        sesh.add(hist)
+    log.info('Created new history entry',
+             extra={'history': hist.id, 'network_name': obs_tuple.network_name,
+                    'lat': obs_tuple.lat, 'lon': obs_tuple.lon})
+    sesh.commit()
+    return hist
+
+
+def get_variable(sesh, network_name, variable_name):
+    log.debug('Get variable from db', extra={'var_name': variable_name,
+                                             'network_name': network_name})
+    variable = sesh.query(Variable).join(Network).filter(and_(
+        Network.name == network_name,
+        Variable.name == variable_name)).first()
+
+    if not variable:
+        log.debug('Unable to match variable')
+        return None
+
+    return variable
+
+
+def get_history(sesh, obs_tuple):
+    log.debug('Find history entry')
+    history = sesh.query(History).join(Station).join(Network).filter(and_(
```

Comment:
This could be multiple history entries. We should name the variable `histories`.